libbpf-rs/src/ringbuf.rs

```diff
@@ -6,82 +6,128 @@ use std::time::Duration;
 
 use crate::*;
 
-/// The canonical interface for managing a collection of [`RingBuffer`]s.
+type Callback = fn(&[u8]) -> i32;
+
+/// Builds [`RingBuffer`] instances.
 ///
 /// `ringbuf`s are a special kind of [`Map`], used to transfer data between
 /// [`Program`]s and userspace.  As of Linux 5.8, the `ringbuf` map is now
 /// preferred over the `perf buffer`.
-pub struct RingBufferManager<F>
-where
-    F: FnMut(&[u8]) -> i32 + 'static,
-{
-    ptr: *mut libbpf_sys::ring_buffer,
-    ringbufs: Vec<Box<F>>,
+#[derive(Default)]
+pub struct RingBufferBuilder {
+    fd_callbacks: Vec<(i32, Box<Callback>)>,
 }
 
-impl<F> RingBufferManager<F>
-where
-    F: FnMut(&[u8]) -> i32 + 'static,
-{
+impl RingBufferBuilder {
+    pub fn new() -> Self {
+        RingBufferBuilder {
+            fd_callbacks: vec![],
+        }
+    }
+
     /// Add a new ringbuf `map` and associated `callback` to this ring buffer
     /// manager. The callback should take one argument, a slice of raw bytes,
     /// and return an i32.
     ///
-    /// Non-zero return values will stop ring buffer consumption early.
+    /// Non-zero return values in the callback will stop ring buffer consumption early.
+    ///
+    /// Callback is defined as:
+    /// ```rust
+    /// type Callback = fn(&[u8]) -> i32;
+    /// ```
     ///
     /// The callback provides a raw byte slice. You may find libraries such as
     /// [`plain`](https://crates.io/crates/plain) helpful.
-    pub fn add_ringbuf(&mut self, map: &Map, callback: F) -> Result<&mut Self> {
-        let callback = Box::new(callback);
+    pub fn add(&mut self, map: &Map, callback: Callback) -> Result<&mut Self> {
```

Comment:
The builder should probably store a reference to `map` to ensure that `map` doesn't go out of scope before `RingBuffer` is built.

Reply:
I'm running into some borrow checker issues with this in the unit tests since `obj.map()` borrows obj as mutable. That's the reason I opted to use map fd rather than storing references to the maps.

Reply:
Right, but the issue I'm worried about is that the owning `Object` could drop before `RingBufferBuilder::build()` is called, dropping the maps and invalidating the FDs before the kernel can increment the refcount on the map.

Maybe we could drop support for the multi-ring manager feature? Would that allow us to hold a ref to `Map` in `RingBufferBuilder`?

Reply:
Multi-ring manager is important, we shouldn't remove the support for it. If someone drops Object before instantiating RingBuffer, that's too bad for them, they are misusing the API. If we can prevent that -- that's great, if not, then it's fine, doesn't seem like a terrible problem to be concerned about.

Reply:
Agreed that multi ring buffer is a killer feature of ringbuf over perf buffers. We should definitely support it if possible.